cypress-io/cypress
# feat: add experimentalModifyObstructiveThirdPartyCode flag for regex rewriter

- Closes #21476 #21307 with the completion of #22479

### User facing changelog

Introduces a new experimental flag, called `experimentalModifyObstructiveThirdPartyCode`. When enabled, `experimentalModifyObstructiveThirdPartyCode` will turn on additional [modifyObstructiveCode](https://docs.cypress.io/guides/references/configuration#modifyObstructiveCode) options to prevent frame busting, as well as strips integrity tags out of `<link>` and `<script>` elements. In the current state of this experimental flag, [SRI](https://developer.mozilla.org/en-US/docs/Web/Security/Subresource_Integrity) is not supported.

### Additional details

As a team, we felt it be best that changes to the `regex-rewriter` for `modifyObstructiveCode` be wrapped in its own experimental flag to reduce the potential impact of the rewriter modify code that breaks users who are not leveraging `cy.origin` or wish to preserve SRI within their application. The impact of this flag will change as we continue to modify the `regex-rewritter` to handle different websites/authentication platforms. Additionally, the team plans to support server-side integrity matching to verify `<link>` and `<script>` elements that have had their integrity stripped be revalidated on the server through hash comparison.

* When the experimental flag is enabled:
  * third party html/scripts are run through  the rewriter
  * regex rewriter will look for patterns similar to `e.top === e.self`
  * subresource integrity hashes removed from script and link tags
  * calling `formElement.submit()` correctly handles unsupported targets

### Steps to test

* additional unit tests to the `regex-rewriter` have been added to help guarantee the expected rewriting behavior for newly added modifications to the strip/stripStream function, as well as adding the `integrity.cy.ts` test inside the `origin` folder. For this test suite to run correctly, both the `experimentalSessionAndOrigin` and `experimentalExpandedModifyObstructiveCode` MUST be set to true. This has been modified in the package.json to pass in CI.

### How has the user experience changed?

Users should now be able to log into Microsoft Online SSO and Login Live SSO (not pictured) via Cypress
![microsoft-login](https://user-images.githubusercontent.com/3980464/176259469-89dd1427-c018-4e4b-bfaf-905bcf8670ea.gif)

### PR Tasks
